pkg/proxy/proxy.go

```diff
@@ -47,29 +49,55 @@ type kubeRBACProxy struct {
 	authorizerAttributesGetter *krpAuthorizerAttributesGetter
 	// config for kube-rbac-proxy
 	Config Config
+	// StaleCache for caching auth requests
+	StaleCache    simpleCache
+	StaleCacheTTL time.Duration
 }
 
-func new(authenticator authenticator.Request, authorizer authorizer.Authorizer, config Config) *kubeRBACProxy {
-	return &kubeRBACProxy{authenticator, authorizer, newKubeRBACProxyAuthorizerAttributesGetter(config.Authorization), config}
+func new(authenticator authenticator.Request, authorizer authorizer.Authorizer, config Config, staleCacheTTL time.Duration) *kubeRBACProxy {
+	proxy := kubeRBACProxy{
+		Request:                    authenticator,
+		Authorizer:                 authorizer,
+		authorizerAttributesGetter: newKubeRBACProxyAuthorizerAttributesGetter(config.Authorization),
+		Config:                     config,
+		StaleCache:                 FakeCache{},
+	}
+	if staleCacheTTL > 0*time.Second {
+		proxy.StaleCache = utilcache.NewLRUExpireCache(4096)
+		proxy.StaleCacheTTL = staleCacheTTL
+	}
+	return &proxy
 }
 
 // New creates an authenticator, an authorizer, and a matching authorizer attributes getter compatible with the kube-rbac-proxy
-func New(client clientset.Interface, config Config, authorizer authorizer.Authorizer, authenticator authenticator.Request) (*kubeRBACProxy, error) {
-	return new(authenticator, authorizer, config), nil
+func New(client clientset.Interface, config Config, authorizer authorizer.Authorizer, authenticator authenticator.Request, staleCacheTTL time.Duration) (*kubeRBACProxy, error) {
+	return new(authenticator, authorizer, config, staleCacheTTL), nil
 }
 
 // Handle authenticates the client and authorizes the request.
 // If the authn fails, a 401 error is returned. If the authz fails, a 403 error is returned
 func (h *kubeRBACProxy) Handle(w http.ResponseWriter, req *http.Request) bool {
+	token := getTokenFromRequest(req)
+	if token == "" {
+		klog.Errorf("Unable to get token from request")
+		http.Error(w, "Unauthorized", http.StatusUnauthorized)
+		return false
+	}
 	// Authenticate
 	u, ok, err := h.AuthenticateRequest(req)
```

Comment:
Wouldn't we instead want to attempt to read the previous response from the cache and only if stale or not existant move forward with the real request? The current strategy looks like we're always going to attempt the real request. Am I missing something?

Reply:
In my thoughts, we have to use this cache only in case of kube-apiserver unconscious.

For example:
If you haven't high available control-plane (or, as people are calling it, multi-master cluster), you will have holes in graphs every time kube-apiserver will go down. 

Prometheus requests metrics periodically. If Prometheus was allowed to scrape targets when kube-apiserver was healthy, we would like to allow him to scrape targets for some period (like 15min or 30min) even if kube-apiserver is not responding.

However, if kube-apiserver responds, we would like to get the most up to date information about the identity and access rights of the user.

Please correct me if I am wrong.

Reply:
Hm, yes. I had the initial assumption as @brancz had, when look at your PR title and I was super excited :smile: 

After the comment, I understand your use case, but you mention this clearly in the README.md.